File: FruitWholesale.Client/src/app/features/purchase/purchase.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { PaginatedList, PaginationRequest } from '../../core/models/common.model';
import { Purchase, PurchaseListItem, SavePurchase } from '../../core/models/transactions.model';
import { toHttpParams } from '../../core/utils/http-params.util';

@Injectable({ providedIn: 'root' })
export class PurchaseService {
  private readonly baseUrl = `${environment.apiUrl}/purchase`;

  constructor(private readonly http: HttpClient) {}

  getPaged(request: PaginationRequest, supplierId?: number | null, fromDate?: string | null, toDate?: string | null) {
    return this.http.get<PaginatedList<PurchaseListItem>>(this.baseUrl, {
      params: toHttpParams({ ...request, supplierId, fromDate, toDate })
    });
  }

  getById(id: number): Observable<Purchase> {
    return this.http.get<Purchase>(`${this.baseUrl}/${id}`);
  }

  getNextInvoiceNo(): Observable<string> {
    return this.http.get(`${this.baseUrl}/next-invoice-no`, { responseType: 'text' });
  }

  create(dto: SavePurchase): Observable<Purchase> {
    return this.http.post<Purchase>(this.baseUrl, dto);
  }

  update(id: number, dto: SavePurchase): Observable<Purchase> {
    return this.http.put<Purchase>(`${this.baseUrl}/${id}`, dto);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}

File: FruitWholesale.Mobile/lib/features/purchase/purchase_service.dart
import '../../core/api/api_client.dart';
import '../../core/models/paginated_list.dart';
import 'purchase_models.dart';

class PurchaseService {
  final ApiClient _api;

  PurchaseService(this._api);

  Future<PaginatedList<PurchaseListItem>> getPaged({
    int pageNumber = 1,
    int pageSize = 20,
    String? searchTerm,
    int? supplierId,
    String? fromDate,
    String? toDate,
  }) async {
    final json = await _api.get('/purchase', query: {
      'pageNumber': pageNumber,
      'pageSize': pageSize,
      if (searchTerm != null && searchTerm.isNotEmpty) 'searchTerm': searchTerm,
      if (supplierId != null) 'supplierId': supplierId,
      if (fromDate != null) 'fromDate': fromDate,
      if (toDate != null) 'toDate': toDate,
    });
    return PaginatedList.fromJson(json as Map<String, dynamic>, PurchaseListItem.fromJson);
  }

  Future<PurchaseDetail> getById(int id) async {
    final json = await _api.get('/purchase/$id');
    return PurchaseDetail.fromJson(json as Map<String, dynamic>);
  }

  Future<String> getNextInvoiceNo() async {
    final json = await _api.get('/purchase/next-invoice-no');
    return json as String;
  }

  Future<PurchaseDetail> create(PurchaseDetail purchase) async {
    final json = await _api.post('/purchase', body: purchase.toSaveJson());
    return PurchaseDetail.fromJson(json as Map<String, dynamic>);
  }

  Future<PurchaseDetail> update(int id, PurchaseDetail purchase) async {
    final json = await _api.put('/purchase/$id', body: purchase.toSaveJson());
    return PurchaseDetail.fromJson(json as Map<String, dynamic>);
  }
}
